### PR TITLE
CON-1154 Added import learner details request for updating ilr

### PR DESCRIPTION
**Pull request by ChrisJWoodcock**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Api.Types/Models/Staff/LearnerDetailForStaff.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 namespace SFA.DAS.AssessorService.Api.Types.Models
 {
-    public class LearnerDetail
+    public class LearnerDetailForStaff
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Api.Types/Models/Staff/LearnerDetailForStaff.cs]: The assessor service does not call the 'Staff' app - so it does not require knowledge of it; my intention here was to give a purpose to a collection of LearnerDetails and other details - I will think of a better name.

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsController.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                return Ok(await _mediator.Send(request));
+            }
+            catch (BadRequestException ex)
+            {
+                _logger.LogError(ex.Message);
+                return BadRequest();
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($@"Bad request, Message: [{ex.Message}]");
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsController.cs]: I'll correct this

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs [hunk 3]
```diff
@@ -1,31 +1,30 @@
...
         [HttpGet]
-        public async Task<IActionResult> Get(int stdCode, 
+        public async Task<IActionResult> Get(int stdCode,
             long uln,
-            bool allLogs=false)
+            bool allLogs = false)
         {
-            return Ok(await _mediator.Send(new LearnerDetailRequest(stdCode,uln,
+            return Ok(await _mediator.Send(new LearnerDetailForStaffRequest(stdCode, uln,
                 allLogs)));
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs]: Auto formatted will correct

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
 
-        public LearnerDetailsController(IMediator mediator)
+        public LearnerDetailsQueryController(IMediator mediator)
         {
             _mediator = mediator;
         }
 
         [HttpGet]
-        public async Task<IActionResult> Get(int stdCode, 
+        public async Task<IActionResult> Get(int stdCode,
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs]: Will correct

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/Learner/LearnerDetailsQueryController.cs]: Auto Formated

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_and_no_existing_learner.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        [Test]
+        public async Task Then_a_new_learner_record_is_created()
+        {
+            // Act
+            Response = await Sut.Handle(Request, new CancellationToken());
+
+            // Assert
+            IlrRepository.Verify(r => r.Create(It.IsAny<string>(), It.IsAny<long>(), Request.Uln.Value, Request.StdCode.Value, It.IsAny<int?>(),
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_and_no_existing_learner.cs]: I see - I'll make some changes to use a model and a simpler verify

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_with_certificate.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        [Test]
+        public async Task Then_an_existing_learner_record_is_replaced()
+        {
+            // Act
+            Response = await Sut.Handle(Request, new CancellationToken());
+
+            // Assert
+            IlrRepository.Verify(r => r.Update(Request.Source, Request.Ukprn.Value, Request.Uln.Value, Request.StdCode.Value,
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_with_certificate.cs]: As above will update

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_previous_source.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        [Test]
+        public async Task Then_learner_records_are_unchanged()
+        {
+            // Act
+            Response = await Sut.Handle(Request, new CancellationToken());
+
+            // Assert
+            IlrRepository.Verify(r => r.Create(It.IsAny<string>(), It.IsAny<long>(), It.IsAny<long>(), It.IsAny<int>(), It.IsAny<int?>(),
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_previous_source.cs]: Will change

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_same_source_and_different_ukprn_with_certificate.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            // Act
+            Response = await Sut.Handle(Request, new CancellationToken());
+
+            // Assert
+            IlrRepository.Verify(r => r.Create(It.IsAny<string>(), It.IsAny<long>(), It.IsAny<long>(), It.IsAny<int>(), It.IsAny<int?>(),
+                It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>(), It.IsAny<DateTime?>(), It.IsAny<DateTime?>(), It.IsAny<int?>(), It.IsAny<string>(),
+                It.IsAny<string>(), It.IsAny<DateTime?>(), It.IsAny<int?>(), It.IsAny<int?>(), It.IsAny<DateTime?>(), It.IsAny<string>()), Times.Never);
+
+            // Assert
+            IlrRepository.Verify(r => r.Update(It.IsAny<string>(), It.IsAny<long>(), It.IsAny<long>(), It.IsAny<int>(), It.IsAny<int?>(),
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_same_source_and_different_ukprn_with_certificate.cs]: Will change

File: src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+            var learner = await _ilrRepository.Get(request.Uln.Value, request.StdCode.Value);
+
+            if (learner == null)
+            {
+                return CreateIlrRecord(request);
+            }
+            else
+            {
+                if (int.Parse(request.Source) < int.Parse(learner.Source))
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs]: No it's not necessary as they represent years such as 1920 2021 2122 and compared as string they will still return the same comparison result - I'll change it

File: src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            return "CreatedLearnerDetail";
+        }
+
+        private string UpdateIlrRecord(ImportLearnerDetailRequest request, bool isUpdate, Ilr currentLearner = null)
+        {
+            _logger.LogInformation("Handling Import Learner Detail Request - Update Ilr");
+
+            // for an update to certain fields if the request is null then the currrent value will 
+            // retained, otherwise the request value will be used
+            _ilrRepository.Update(request.Source, request.Ukprn.Value, request.Uln.Value, request.StdCode.Value, request.FundingModel, request.GivenNames, request.FamilyName,
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs]: I'll refactor to use a method to make it more readable

File: src/SFA.DAS.AssessorService.Application/Interfaces/IIlrRepository.cs [hunk 11]
```diff
@@ -7,8 +8,17 @@ namespace SFA.DAS.AssessorService.Application.Interfaces
     public interface IIlrRepository
     {
         Task<IEnumerable<Ilr>> SearchForLearnerByUln(long uln);
+        
         Task<Ilr> Get(long uln, int standardCode);
+        
         Task StoreSearchLog(SearchLog log);
-        Task<IEnumerable<Ilr>> Search(string searchQuery);
+        
+        Task Create(string source, long ukprn, long uln, int stdCode, int? fundingModel, string givenNames, string familyName,
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application/Interfaces/IIlrRepository.cs]: I'll refactor to use a model there is no good reason not too

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_without_certificate.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        [SetUp]
+        public void Arrange()
+        {
+            BaseArrange();
+
+            Request = CreateImportLearnerDetailRequest(LearnerTwo);
+
+            Request.Source = "2021";
+            Request.EpaOrgId = null; // replacement null value
```
Comment (cofaulco) [inline src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_without_certificate.cs]: a single comment would probably have been sufficient 😉

File: src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+            _logger.LogInformation("Handling Import Learner Detail Request");
+
+            return CheckMissingMandatoryFields(request) ?? new ImportLearnerDetailResponse
+            {
+                Result = await HandleRequest(request, cancellationToken)
+            };
+        }
+
+        private async Task<string> HandleRequest(ImportLearnerDetailRequest request, CancellationToken cancellationToken)
+        {
```
Comment (cofaulco) [inline src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs]: would be nice to simplify this code, but might be one for a tech debt task

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_same_source_and_different_ukprn_with_certificate.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+            // Assert
+            IlrRepository.Verify(r => r.Create(It.IsAny<Ilr>()), Times.Never);
+        }
+
+        [Test]
+        public async Task Then_learner_records_are_not_updated()
+        {
+            // Arrange
+            Request = CreateImportLearnerDetailRequest(LearnerThree.Source, 444444444, LearnerThree.Uln, LearnerThree.StdCode,
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_same_source_and_different_ukprn_with_certificate.cs]: I was using an existing one with some difference values

File: src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_without_certificate.cs (hunk 12 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application.UnitTests/Handlers/Learner/When_called_for_existing_learner_having_later_source_without_certificate.cs]: Corrected

File: src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs (hunk 13 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AssessorService.Application/Handlers/Learner/ImportLearnerDetailHandler.cs]: Updated to reduce nesting and removed duplicated return values.

**Pull request merged**